runtime/compiler/optimizer/J9Inliner.cpp

```diff
@@ -552,6 +552,29 @@ bool TR_J9VirtualCallSite::findCallSiteTarget(TR_CallStack *callStack, TR_Inline
 
    tryToRefineReceiverClassBasedOnResolvedTypeArgInfo(inliner);
 
+   // Refine receiver class based on CP class
+   // When we have an invokevirtual on an abstract method defined in an interface class,
+   // the call site's class will be more concrete than class of method.
+   // This happens when an abstract class implements an interface class without providing
+   // implementation for the given method, and the call site is refering to the method of
+   // the abstract class, the cp entry of the method ref will be resolved to j9method of
+   // the interface class. However, the class ref from cp will be resolved to the abstract
+   // class, which is more concrete
+   //
+   if (_cpIndex != -1 && _receiverClass && TR::Compiler->cls.isInterfaceClass(comp(), _receiverClass))
+      {
+      TR_ResolvedMethod* owningMethod = _initialCalleeMethod->owningMethod();
+      int32_t classRefCPIndex = owningMethod->classCPIndexOfMethod(_cpIndex);
+      TR_OpaqueClassBlock* callSiteClass = owningMethod->getClassFromConstantPool(comp(), classRefCPIndex);
+      if (callSiteClass &&
+          callSiteClass != _receiverClass &&
+          fe()->isInstanceOf(callSiteClass, _receiverClass, true, true, false) == TR_yes)
+         {
+         heuristicTrace(inliner->tracer(), "Receiver type is class of an interface method for invokevirtual, improve it to call site receiver type\n");
```

Comment:
Should you output the new refined receiver class address that was found as part of this trace message?  Outputting which invokevirtual it was would also be useful.

Reply:
@0xdaryl Added change to print out the old and new class info